testgrid/cmd/summarizer/summarizer.go

```diff
@@ -0,0 +1,116 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package summarizer converts input table data aggregated from test results
+// into the structured DashboardTabSummary proto for Testgrid consumption.
+package summarizer
+
+import (
+	"fmt"
+	"net/url"
+
+	configpb "k8s.io/test-infra/testgrid/config"
+	summarypb "k8s.io/test-infra/testgrid/summary"
+)
+
+// TableToSummary converts input queryData collected from test results to
+// DashboardTabSummary proto.
+func TableToSummary(queryData Response, dashboardName string, dashboardTab configpb.DashboardTab, updateTimestamp int64) (summarypb.DashboardTabSummary, error) {
+	s := extractBasicSummary(queryData, dashboardName, dashboardTab, updateTimestamp)
+	s.FailingTestSummaries = extractTests(queryData)
+	return s, nil
+}
+
+func extractBasicSummary(queryData Response, dashboardName string, dashboardTab configpb.DashboardTab, updateTimestamp int64) summarypb.DashboardTabSummary {
+	var bugURL string
+	if len(queryData.OpenBugTemplate.URL) > 0 {
+		bugURL = queryData.OpenBugTemplate.URL
+	}
+	lastRunTimestamp := 0
+	for i, e := range queryData.Timestamps {
+		if i == 0 || e > lastRunTimestamp {
+			lastRunTimestamp = e
+		}
+	}
+	overallStatus := summarypb.DashboardTabSummary_UNKNOWN
+	if queryData.OverallStatus > 0 {
+		overallStatus = queryData.OverallStatus
+	}
+
+	s := summarypb.DashboardTabSummary{
+		DashboardName:       dashboardName,
+		DashboardTabName:    dashboardTab.Name,
+		Alert:               queryData.Alerts,
+		LastUpdateTimestamp: float64(updateTimestamp),
+		LastRunTimestamp:    float64(lastRunTimestamp),
+		LatestGreen:         queryData.LatestGreen,
+		OverallStatus:       overallStatus,
+		Status:              queryData.Summary,
+		BugUrl:              bugURL,
+	}
+
+	return s
+}
+
+func appendQuery(query, newKey, newVal string) string {
+	if len(query) > 0 {
+		return fmt.Sprintf("%s&%s=%s", query, newKey, newVal)
+	} else {
+		return fmt.Sprintf("%s=%s", newKey, newVal)
+	}
+}
+
+func extractTests(queryData Response) []*summarypb.FailingTestSummary {
+	var failingTestSummaries []*summarypb.FailingTestSummary
+	if len(queryData.Tests) > 0 {
```

Comment:
don't need this if block, you'll get this for free from the `range` below; if it's an empty list, there's nothing to iterate over